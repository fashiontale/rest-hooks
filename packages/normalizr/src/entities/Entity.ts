import { NotImplementedError } from '../errors';
import * as schema from '../schema';
import { Schema, AbstractInstanceType } from '../types';
import SimpleRecord from './SimpleRecord';

/** Represents data that should be deduped by specifying a primary key. */
export default abstract class Entity extends SimpleRecord {
  /**
   * A unique identifier for each Entity
   *
   * @param [parent] When normalizing, the object which included the entity
   * @param [key] When normalizing, the key where this entity was found
   */
  abstract pk(parent?: any, key?: string): string | undefined;

  /** Returns the globally unique identifier for the static Entity */
  static get key(): string {
    /* istanbul ignore next */
    if (process.env.NODE_ENV !== 'production' && this.name === '')
      throw new Error(
        'Entity classes without a name must define `static get key()`',
      );
    return this.name;
  }

  /** Defines nested entities */
  static schema: { [k: string]: Schema } = {};

  /** Defines indexes to enable lookup by */
  declare static indexes?: readonly string[];

  /**
   * A unique identifier for each Entity
   *
   * @param [value] POJO of the entity or subset used
   * @param [parent] When normalizing, the object which included the entity
   * @param [key] When normalizing, the key where this entity was found
   */
  static pk<T extends typeof Entity>(
    this: T,
    value: Partial<AbstractInstanceType<T>>,
    parent?: any,
    key?: string,
  ): string | undefined {
    return this.prototype.pk.call(value, parent, key) || key;
  }

  /** Returns this to be used in a schema definition.
   * This is essential to capture the correct type to be used in inferencing.
   */
  static asSchema<T extends typeof Entity>(this: T) {
    return this as EntitySchema<T>;
  }

  static normalize(
    input: Partial<Entity>,
    parent: any,
    key: string | undefined,
    visit: Function,
    addEntity: Function,
    visitedEntities: Record<string, any>,
  ) {
    // TODO: what's store needs to be a differing type from fromJS
    const processedEntity = this.fromJS(input, parent, key);
    const id = processedEntity.pk(parent, key);
    /* istanbul ignore next */
    if (id === undefined) {
      if (process.env.NODE_ENV !== 'production' && id === undefined) {
        throw new Error(
          `Missing usable resource key when normalizing response.

  This is likely due to a malformed response.
  Try inspecting the network response or fetch() return value.

  Entity: ${this}
  Value: ${input && JSON.stringify(input, null, 2)}
  `,
        );
      } else {
        throw new Error('undefined pk');
      }
    }
    const entityType = this.key;

    if (!(entityType in visitedEntities)) {
      visitedEntities[entityType] = {};
    }
    if (!(id in visitedEntities[entityType])) {
      visitedEntities[entityType][id] = [];
    }
    if (
      visitedEntities[entityType][id].some((entity: any) => entity === input)
    ) {
      return id;
    }
    visitedEntities[entityType][id].push(input);

    Object.keys(this.schema).forEach(key => {
      if (
        Object.hasOwnProperty.call(processedEntity, key) &&
        typeof processedEntity[key] === 'object'
      ) {
        const schema = this.schema[key];
        processedEntity[key] = visit(
          processedEntity[key],
          processedEntity,
          key,
          schema,
          addEntity,
          visitedEntities,
        );
      }
    });

    addEntity(this, processedEntity, input, parent, key);
    return id;
  }

  // TODO: Add denormalizing capability
  static denormalize<T extends typeof Entity>(
    this: T,
    entity: AbstractInstanceType<T> | undefined,
    unvisit: schema.UnvisitFunction,
  ): [AbstractInstanceType<T>, true] {
    return [entity, true] as any;
  }
}

/* istanbul ignore next */
if (process.env.NODE_ENV !== 'production') {
  // for those not using TypeScript this is a good catch to ensure they are defining
  // the abstract members
  Entity.fromJS = function fromJS<T extends typeof SimpleRecord>(
    this: T,
    props: Partial<AbstractInstanceType<T>>,
  ): AbstractInstanceType<T> {
    if ((this as any).prototype.pk === undefined)
      throw new Error('cannot construct on abstract types');
    return SimpleRecord.fromJS.call(this, props) as any;
  };
}

export type EntitySchema<E extends typeof Entity> = E & {
  normalize(
    input: any,
    parent: any,
    key: any,
    visit: Function,
    addEntity: Function,
    visitedEntities: Record<string, any>,
  ): string;
  denormalize(
    entity: any,
    unvisit: Function,
  ): [AbstractInstanceType<E>, boolean];
  _normalizeNullable(): string | undefined;
  _denormalizeNullable(): [AbstractInstanceType<E> | undefined, boolean];
};

export function isEntity(schema: Schema | null): schema is typeof Entity {
  return schema !== null && (schema as any).pk !== undefined;
}
