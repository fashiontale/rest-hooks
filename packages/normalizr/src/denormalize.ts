import * as ImmutableUtils from './schemas/ImmutableUtils';
import * as ArrayUtils from './schemas/Array';
import * as ObjectUtils from './schemas/Object';
import { Denormalize, DenormalizeNullable, Schema } from './types';
import SimpleRecord from './entities/SimpleRecord';
import { isEntity } from './entities/Entity';

const unvisitEntity = (
  id: any,
  schema: any,
  unvisit: any,
  getEntity: any,
  cache: Record<string, any>,
): [any, boolean] => {
  const entity = getEntity(id, schema);
  if (typeof entity !== 'object' || entity === null) {
    return [entity, false];
  }

  if (!cache[schema.key]) {
    cache[schema.key] = {};
  }

  let found = true;
  if (!cache[schema.key][id]) {
    // Ensure we don't mutate it non-immutable objects
    const entityCopy =
      ImmutableUtils.isImmutable(entity) || entity instanceof SimpleRecord
        ? entity
        : { ...entity };

    // Need to set this first so that if it is referenced further within the
    // denormalization the reference will already exist.
    cache[schema.key][id] = entityCopy;
    [cache[schema.key][id], found] = schema.denormalize(entityCopy, unvisit);
  }

  return [cache[schema.key][id], found];
};

const getUnvisit = (entities: Record<string, any>) => {
  const cache = {};
  const getEntity = getEntities(entities);

  return [
    function unvisit(input: any, schema: any): [any, boolean] {
      if (!schema) return [input, true];

      if (
        typeof schema === 'object' &&
        (!schema.denormalize || typeof schema.denormalize !== 'function')
      ) {
        const method = Array.isArray(schema)
          ? ArrayUtils.denormalize
          : ObjectUtils.denormalize;
        return method(schema, input, unvisit);
      }

      // null is considered intentional, thus always 'found' as true
      if (input === null) {
        return [input, true];
      }

      if (isEntity(schema)) {
        // unvisitEntity just can't handle undefined
        if (input === undefined) {
          return [input, false];
        }
        return unvisitEntity(input, schema, unvisit, getEntity, cache);
      }

      if (typeof schema.denormalize === 'function') {
        return schema.denormalize(input, unvisit);
      }

      return [input, true];
    },
    cache,
  ] as const;
};

const getEntities = (entities: Record<string, any>) => {
  const isImmutable = ImmutableUtils.isImmutable(entities);

  return (entityOrId: any, schema: any) => {
    const schemaKey = schema.key;

    if (typeof entityOrId === 'object') {
      return entityOrId;
    }

    if (isImmutable) {
      return entities.getIn([schemaKey, entityOrId.toString()]);
    }

    return entities[schemaKey] && entities[schemaKey][entityOrId];
  };
};

export const denormalize = <S extends Schema>(
  input: any,
  schema: S,
  entities: any,
):
  | [Denormalize<S>, true, Record<string, Record<string, any>>]
  | [DenormalizeNullable<S>, false, Record<string, Record<string, any>>] => {
  /* istanbul ignore next */
  // eslint-disable-next-line no-undef
  if (process.env.NODE_ENV !== 'production' && schema === undefined)
    throw new Error('shema needed');
  if (typeof input !== 'undefined') {
    const [unvisit, cache] = getUnvisit(entities);
    return [...unvisit(input, schema), cache] as any;
  }
  return [undefined, false, {}] as any;
};

export const denormalizeSimple = <S extends Schema>(
  input: any,
  schema: S,
  entities: any,
) => denormalize(input, schema, entities).slice(0, 2);
