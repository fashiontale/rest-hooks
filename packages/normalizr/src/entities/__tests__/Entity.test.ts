// eslint-env jest
import { fromJS, Record } from 'immutable';
import { first } from 'lodash';

import { denormalizeSimple as denormalize } from '../../denormalize';
import { normalize, schema, SimpleRecord, AbstractInstanceType } from '../../';
import Entity from '../Entity';
import IDEntity from '../IDEntity';

const values = obj => Object.keys(obj).map(key => obj[key]);

class Tacos extends IDEntity {
  readonly name: string = '';
  readonly alias: string | undefined = undefined;
}

describe(`${Entity.name} normalization`, () => {
  test('normalizes an entity', () => {
    class MyEntity extends IDEntity {}
    expect(normalize({ id: 1 }, MyEntity)).toMatchSnapshot();
  });

  describe('key', () => {
    test('must be created with a key name', () => {
      const makeSchema = () =>
        class extends Entity {
          readonly id: number = 0;
          pk() {
            return `${this.id}`;
          }
        };
      expect(() => makeSchema().key).toThrow();
    });

    test('key name must be a string', () => {
      // @ts-expect-error
      class MyEntity extends IDEntity {
        static get key() {
          return 42;
        }
      }
    });

    test('key getter should return key set via `static get key()`', () => {
      class User extends IDEntity {
        static get key() {
          return 'user';
        }
      }
      expect(User.key).toEqual('user');
    });
  });

  describe('idAttribute', () => {
    test('can use a custom idAttribute string', () => {
      class User extends Entity {
        readonly idStr: string = '';
        readonly name: string = '';

        pk() {
          return this.idStr;
        }
      }
      expect(
        normalize({ idStr: '134351', name: 'Kathy' }, User.asSchema()),
      ).toMatchSnapshot();
    });

    test('can normalize entity IDs based on their object key', () => {
      class User extends Entity {
        readonly name: string = '';
        pk(parent?: any, key?: string) {
          return key;
        }
      }
      const inputSchema = new schema.Values(
        { users: User.asSchema() },
        () => 'users',
      );

      expect(
        normalize(
          { 4: { name: 'taco' }, 56: { name: 'burrito' } },
          inputSchema,
        ),
      ).toMatchSnapshot();
    });

    test("can build the entity's ID from the parent object", () => {
      class User extends Entity {
        readonly id: string = '';
        readonly name: string = '';
        pk(parent?: any, key?: string) {
          return `${parent.name}-${key}-${this.id}`;
        }
      }
      const inputSchema = new schema.Object({ user: User.asSchema() });

      expect(
        normalize(
          { name: 'tacos', user: { id: '4', name: 'Jimmy' } },
          inputSchema,
        ),
      ).toMatchSnapshot();
    });
  });

  describe('mergeStrategy', () => {
    test('defaults to plain merging', () => {
      expect(
        normalize(
          [
            { id: 1, name: 'foo' },
            { id: 1, name: 'bar', alias: 'bar' },
          ],
          [Tacos.asSchema()],
        ),
      ).toMatchSnapshot();
    });

    test('can use a custom merging strategy', () => {
      class MergeTaco extends Tacos {
        static merge<T extends typeof SimpleRecord>(
          this: T,
          first: AbstractInstanceType<T>,
          second: AbstractInstanceType<T>,
        ) {
          const props = Object.assign(
            {},
            this.toObjectDefined(first),
            this.toObjectDefined(second),
            { name: (first as MergeTaco).name },
          );
          return this.fromJS(props);
        }
      }

      expect(
        normalize(
          [
            { id: 1, name: 'foo' },
            { id: 1, name: 'bar', alias: 'bar' },
          ],
          [MergeTaco.asSchema()],
        ),
      ).toMatchSnapshot();
    });
  });

  describe('processStrategy', () => {
    test('can use a custom processing strategy', () => {
      class ProcessTaco extends Tacos {
        static fromJS<T extends typeof SimpleRecord>(
          this: T,
          props: Partial<AbstractInstanceType<T>>,
          parent?: any,
          key?: string,
        ) {
          return super.fromJS({
            ...props,
            slug: `thing-${(props as ProcessTaco).id}`,
          }) as AbstractInstanceType<T>;
        }
      }

      expect(
        normalize({ id: 1, name: 'foo' }, ProcessTaco.asSchema()),
      ).toMatchSnapshot();
    });

    test('can use information from the parent in the process strategy', () => {
      class ChildEntity extends IDEntity {
        readonly content: string = '';

        static fromJS<T extends typeof SimpleRecord>(
          this: T,
          props: Partial<AbstractInstanceType<T>>,
          parent?: any,
          key?: string,
        ) {
          return super.fromJS({
            ...props,
            parentId: parent.id,
            parentKey: key,
          }) as AbstractInstanceType<T>;
        }
      }
      class ParentEntity extends IDEntity {
        readonly content: string = '';

        static schema = { child: ChildEntity.asSchema() };
      }

      expect(
        normalize(
          {
            id: 1,
            content: 'parent',
            child: { id: 4, content: 'child' },
          },
          ParentEntity.asSchema(),
        ),
      ).toMatchSnapshot();
    });

    test('is run before and passed to the schema normalization', () => {
      class AttachmentsEntity extends IDEntity {}
      class EntriesEntity extends IDEntity {
        static schema = {
          data: { attachment: AttachmentsEntity.asSchema() },
        };

        static fromJS<T extends typeof SimpleRecord>(
          this: T,
          props: Partial<AbstractInstanceType<T>>,
          parent?: any,
          key?: string,
        ) {
          return super.fromJS({
            ...values(props)[0],
            type: Object.keys(props)[0],
          }) as AbstractInstanceType<T>;
        }
      }

      expect(
        normalize(
          { message: { id: '123', data: { attachment: { id: '456' } } } },
          EntriesEntity.asSchema(),
        ),
      ).toMatchSnapshot();
    });
  });
});

/* TODO: This isn't supported quite yet
describe(`${Entity.name} denormalization`, () => {
  test('denormalizes an entity', () => {
    const entities = {
      tacos: {
        1: { id: 1, type: 'foo' },
      },
    };
    expect(denormalize(1, Tacos.asSchema(), entities)).toMatchSnapshot();
    expect(
      denormalize(1, Tacos.asSchema(), fromJS(entities)),
    ).toMatchSnapshot();
  });

  class Food extends IDEntity {}
  class Menu extends IDEntity {
    readonly food: Food = Food.fromJS();

    static schema = { food: Food.asSchema() };
  }

  test('denormalizes deep entities', () => {
    const entities = {
      menus: {
        1: { id: 1, food: 1 },
        2: { id: 2 },
      },
      foods: {
        1: { id: 1 },
      },
    };

    expect(denormalize(1, Menu.asSchema(), entities)).toMatchSnapshot();
    expect(denormalize(1, Menu.asSchema(), fromJS(entities))).toMatchSnapshot();

    expect(denormalize(2, Menu.asSchema(), entities)).toMatchSnapshot();
    expect(denormalize(2, Menu.asSchema(), fromJS(entities))).toMatchSnapshot();
  });

  test('denormalizes to undefined for missing data', () => {
    const entities = {
      menus: {
        1: { id: 1, food: 2 },
      },
      foods: {
        1: { id: 1 },
      },
    };

    expect(denormalize(1, Menu.asSchema(), entities)).toMatchSnapshot();
    expect(denormalize(1, Menu.asSchema(), fromJS(entities))).toMatchSnapshot();

    expect(denormalize(2, Menu.asSchema(), entities)).toMatchSnapshot();
    expect(denormalize(2, Menu.asSchema(), fromJS(entities))).toMatchSnapshot();
  });

  test('denormalizes deep entities with records', () => {
    const Food = Record({ id: null });
    const MenuR = Record({ id: null, food: null });

    const entities = {
      menus: {
        1: new MenuR({ id: 1, food: 1 }),
        2: new MenuR({ id: 2 }),
      },
      foods: {
        1: new Food({ id: 1 }),
      },
    };

    expect(denormalize(1, Menu.asSchema(), entities)).toMatchSnapshot();
    expect(denormalize(1, Menu.asSchema(), fromJS(entities))).toMatchSnapshot();

    expect(denormalize(2, Menu.asSchema(), entities)).toMatchSnapshot();
    expect(denormalize(2, Menu.asSchema(), fromJS(entities))).toMatchSnapshot();
  });

  test('can denormalize already partially denormalized data', () => {
    const entities = {
      menus: {
        1: { id: 1, food: { id: 1 } },
      },
      foods: {
        1: { id: 1 },
      },
    };

    expect(denormalize(1, Menu.asSchema(), entities)).toMatchSnapshot();
    expect(denormalize(1, Menu.asSchema(), fromJS(entities))).toMatchSnapshot();
  });

  class User extends IDEntity {
    readonly role = '';
    readonly reports: Report[] = [];
  }
  class Report extends IDEntity {
    readonly title: string = '';
    readonly draftedBy: User = User.fromJS({});
    readonly publishedBy: User = User.fromJS({});

    static schema = {
      draftedBy: User,
      publishedBy: User,
    };
  }
  User.schema = {
    reports: [Report],
  };

  test('denormalizes recursive dependencies', () => {
    const entities = {
      reports: {
        '123': {
          id: '123',
          title: 'Weekly report',
          draftedBy: '456',
          publishedBy: '456',
        },
      },
      users: {
        '456': {
          id: '456',
          role: 'manager',
          reports: ['123'],
        },
      },
    };
    expect(denormalize('123', Report.asSchema(), entities)).toMatchSnapshot();
    expect(
      denormalize('123', Report.asSchema(), fromJS(entities)),
    ).toMatchSnapshot();

    expect(denormalize('456', User.asSchema(), entities)).toMatchSnapshot();
    expect(
      denormalize('456', User.asSchema(), fromJS(entities)),
    ).toMatchSnapshot();
  });

  test('denormalizes entities with referential equality', () => {
    const entities = {
      reports: {
        '123': {
          id: '123',
          title: 'Weekly report',
          draftedBy: '456',
          publishedBy: '456',
        },
      },
      users: {
        '456': {
          id: '456',
          role: 'manager',
          reports: ['123'],
        },
      },
    };

    const [denormalizedReport] = denormalize(
      '123',
      Report.asSchema(),
      entities,
    );

    expect(denormalizedReport).toBe(denormalizedReport.draftedBy.reports[0]);
    expect(denormalizedReport.publishedBy).toBe(denormalizedReport.draftedBy);

    // NOTE: Given how immutable data works, referential equality can't be
    // maintained with nested denormalization.
  });
});
*/
