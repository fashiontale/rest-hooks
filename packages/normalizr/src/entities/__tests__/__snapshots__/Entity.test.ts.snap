// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Entity normalization idAttribute can build the entity's ID from the parent object 1`] = `
Object {
  "entities": Object {
    "User": Object {
      "tacos-user-4": User {
        "id": "4",
        "name": "Jimmy",
      },
    },
  },
  "indexes": Object {},
  "result": Object {
    "name": "tacos",
    "user": "tacos-user-4",
  },
}
`;

exports[`Entity normalization idAttribute can normalize entity IDs based on their object key 1`] = `
Object {
  "entities": Object {
    "User": Object {
      "4": User {
        "name": "taco",
      },
      "56": User {
        "name": "burrito",
      },
    },
  },
  "indexes": Object {},
  "result": Object {
    "4": Object {
      "id": "4",
      "schema": "users",
    },
    "56": Object {
      "id": "56",
      "schema": "users",
    },
  },
}
`;

exports[`Entity normalization idAttribute can use a custom idAttribute string 1`] = `
Object {
  "entities": Object {
    "User": Object {
      "134351": User {
        "idStr": "134351",
        "name": "Kathy",
      },
    },
  },
  "indexes": Object {},
  "result": "134351",
}
`;

exports[`Entity normalization mergeStrategy can use a custom merging strategy 1`] = `
Object {
  "entities": Object {
    "MergeTaco": Object {
      "1": MergeTaco {
        "alias": "bar",
        "id": 1,
        "name": "foo",
      },
    },
  },
  "indexes": Object {},
  "result": Array [
    "1",
    "1",
  ],
}
`;

exports[`Entity normalization mergeStrategy defaults to plain merging 1`] = `
Object {
  "entities": Object {
    "Tacos": Object {
      "1": Tacos {
        "alias": "bar",
        "id": 1,
        "name": "bar",
      },
    },
  },
  "indexes": Object {},
  "result": Array [
    "1",
    "1",
  ],
}
`;

exports[`Entity normalization normalizes an entity 1`] = `
Object {
  "entities": Object {
    "MyEntity": Object {
      "1": MyEntity {
        "id": 1,
      },
    },
  },
  "indexes": Object {},
  "result": "1",
}
`;

exports[`Entity normalization processStrategy can use a custom processing strategy 1`] = `
Object {
  "entities": Object {
    "ProcessTaco": Object {
      "1": ProcessTaco {
        "alias": undefined,
        "id": 1,
        "name": "foo",
        "slug": "thing-1",
      },
    },
  },
  "indexes": Object {},
  "result": "1",
}
`;

exports[`Entity normalization processStrategy can use information from the parent in the process strategy 1`] = `
Object {
  "entities": Object {
    "ChildEntity": Object {
      "4": ChildEntity {
        "content": "child",
        "id": 4,
        "parentId": 1,
        "parentKey": "child",
      },
    },
    "ParentEntity": Object {
      "1": ParentEntity {
        "child": "4",
        "content": "parent",
        "id": 1,
      },
    },
  },
  "indexes": Object {},
  "result": "1",
}
`;

exports[`Entity normalization processStrategy is run before and passed to the schema normalization 1`] = `
Object {
  "entities": Object {
    "AttachmentsEntity": Object {
      "456": AttachmentsEntity {
        "id": "456",
      },
    },
    "EntriesEntity": Object {
      "123": EntriesEntity {
        "data": Object {
          "attachment": "456",
        },
        "id": "123",
        "type": "message",
      },
    },
  },
  "indexes": Object {},
  "result": "123",
}
`;
